File: dat-pypi/src/saro_dat/dat_cms_manager.py
import logging
import threading
import time
import urllib.request
import urllib.error
from typing import Optional, Union

from .dat_manager import DatManager
from .dat import Dat, DatPayload

logger = logging.getLogger(__name__)

class DatCmsManager:
    DAT_CMS_API_VERSION = "v1"

    def __init__(
        self,
        uri: str,
        token: str,
        interval_seconds: int = 60,
        verify_only: bool = False,
        dat_manager: Optional[DatManager] = None
    ):
        self._uri = uri
        self._token = token
        self._interval_seconds = interval_seconds
        self._verify_only = verify_only
        self._manager = dat_manager or DatManager()
        self._version = 0
        self._lock = threading.Lock()
        self._timer: Optional[threading.Timer] = None
        self._stopped = False

        self.sync()

        if self._interval_seconds > 0:
            self._schedule_sync()

    def _schedule_sync(self):
        with self._lock:
            if not self._stopped:
                self._timer = threading.Timer(self._interval_seconds, self._run_sync_task)
                self._timer.daemon = True
                self._timer.start()

    def _run_sync_task(self):
        try:
            self.sync()
        finally:
            self._schedule_sync()

    def stop(self):
        with self._lock:
            self._stopped = True
            if self._timer:
                self._timer.cancel()

    def sync(self):
        if not self._lock.acquire(blocking=False):
            logger.warning("Last request ignored (Duplicate request)")
            return

        try:
            url = f"{self._uri}?version={self._version}"

            request = urllib.request.Request(url)
            request.add_header("Authorization", self._token)
            
            with urllib.request.urlopen(request, timeout=10) as response:
                if response.status != 200:
                    logger.error(f"Response status error, status:{response.status} in {url}")
                    return

                body = response.read().decode('utf-8')
                
                if not body:
                    logger.debug(f"No new certificate: {url}")
                    return

                lines = body.split('\n', 1)
                if len(lines) < 2:
                    if body.startswith('\n'):
                         logger.error(f"Invalid response: {url}")
                         return
                    logger.debug(f"No new certificate: {url}")
                    return

                new_version_str = lines[0].strip()
                new_certificates = lines[1].strip()

                if not new_version_str:
                    logger.error(f"Invalid version in response: {url}")
                    return

                try:
                    new_version = int(new_version_str)
                    renew_count = self._manager.imports(new_certificates, False)
                    self._version = new_version
                    logger.debug(f"Renewed {renew_count} certificates for version {new_version}: {url}")
                except ValueError as e:
                    logger.error(f"Failed to parse version or certificates: {e}")

        except urllib.error.URLError as e:
            logger.error(f"[Exception] DAT CMS Sync {self._uri}: {e}")
        except Exception as e:
            logger.exception(f"[Exception] DAT CMS Sync {self._uri}")
        finally:
            self._lock.release()

    def get_manager(self) -> DatManager:
        return self._manager

    def issue(self, plain: Union[bytes, str, None], secure: Union[bytes, str, None]) -> str:
        return self._manager.issue(plain, secure)

    def parse(self, dat: Union[Dat, str, None]) -> DatPayload:
        return self._manager.parse(dat)

    @classmethod
    def builder(cls):
        return DatCmsManagerBuilder()

class DatCmsManagerBuilder:
    def __init__(self):
        self._uri = "http://localhost:8088"
        self._token = ""
        self._verify_only = False
        self._interval_seconds = 60

    def uri(self, uri: str):
        self._uri = uri.rstrip('/')
        return self

    def token(self, token: str):
        self._token = token
        return self

    def verify_only(self, verify_only: bool):
        self._verify_only = verify_only
        return self

    def interval_seconds(self, interval_seconds: int):
        self._interval_seconds = interval_seconds
        return self

    def interval_off(self):
        return self.interval_seconds(0)

    def build(self) -> DatCmsManager:
        from urllib.parse import urlparse
        parsed = urlparse(self._uri)
        
        if parsed.path and parsed.path != '/':
            raise ValueError(f"uri must be path-less: {self._uri}")
        if parsed.query:
            raise ValueError(f"uri must be query-less: {self._uri}")

        path = "/v1/certs/verify-only" if self._verify_only else "/v1/certs"

        final_uri = f"{parsed.scheme}://{parsed.netloc}{path}"
        
        return DatCmsManager(
            uri=final_uri,
            token=self._token,
            interval_seconds=self._interval_seconds,
            verify_only=self._verify_only
        )

File: dat-ruby/lib/saro/dat/dat_cms_manager.rb
# frozen_string_literal: true

require 'net/http'
require 'uri'
require 'logger'
require 'thread'
require_relative 'dat_manager'
require_relative 'dat'

module Saro
  module Dat
    class DatCmsManager
      DAT_CMS_API_VERSION = "v1"

      def initialize(uri:, token:, interval_seconds: 60, verify_only: false, dat_manager: nil)
        @uri = uri
        @token = token
        @interval_seconds = interval_seconds
        @verify_only = verify_only
        @manager = dat_manager || DatManager.new
        @version = 0
        @lock = Mutex.new
        @stopped = false
        @logger = Logger.new($stdout)
        @logger.level = Logger::DEBUG

        sync

        if @interval_seconds > 0
          schedule_sync
        end
      end

      def stop
        @lock.synchronize do
          @stopped = true
          @thread&.kill # 혹은 다른 방식으로 스레드 중지
        end
      end

      def sync
        # non-blocking lock
        unless @lock.try_lock
          @logger.warn("Last request ignored (Duplicate request)")
          return
        end

        begin
          url = URI("#{@uri}?version=#{@version}")
          request = Net::HTTP::Get.new(url)
          request["Authorization"] = @token

          response = Net::HTTP.start(url.host, url.port, use_ssl: url.scheme == 'https', open_timeout: 10, read_timeout: 10) do |http|
            http.request(request)
          end

          if response.code != "200"
            @logger.error("Response status error, status:#{response.code} in #{url}")
            return
          end

          body = response.body
          if body.nil? || body.empty?
            @logger.debug("No new certificate: #{url}")
            return
          end

          lines = body.split("\n", 2)
          if lines.length < 2
            if body.start_with?("\n")
              @logger.error("Invalid response: #{url}")
              return
            end
            @logger.debug("No new certificate: #{url}")
            return
          end

          new_version_str = lines[0].strip
          new_certificates = lines[1].strip

          if new_version_str.empty?
            @logger.error("Invalid version in response: #{url}")
            return
          end

          begin
            new_version = Integer(new_version_str)
            renew_count = @manager.imports(new_certificates, clear: false)
            @version = new_version
            @logger.debug("Renewed #{renew_count} certificates for version #{new_version}: #{url}")
          rescue ArgumentError => e
            @logger.error("Failed to parse version or certificates: #{e.message}")
          end

        rescue StandardError => e
          @logger.error("[Exception] DAT CMS Sync #{@uri}: #{e.message}")
        ensure
          @lock.unlock
        end
      end

      def get_manager
        @manager
      end

      def issue(plain, secure)
        @manager.issue(plain, secure)
      end

      def parse(dat)
        @manager.parse(dat)
      end

      def self.builder
        DatCmsManagerBuilder.new
      end

      private

      def schedule_sync
        @thread = Thread.new do
          loop do
            sleep(@interval_seconds)
            break if @stopped
            run_sync_task
          end
        end
      end

      def run_sync_task
        sync
      rescue StandardError => e
        @logger.error("Error in sync task: #{e.message}")
      end
    end

    class DatCmsManagerBuilder
      def initialize
        @uri = "http://localhost:8088"
        @token = ""
        @verify_only = false
        @interval_seconds = 60
      end

      def uri(uri)
        @uri = uri.delete_suffix('/')
        self
      end

      def token(token)
        @token = token
        self
      end

      def verify_only(verify_only)
        @verify_only = verify_only
        self
      end

      def interval_seconds(interval_seconds)
        @interval_seconds = interval_seconds
        self
      end

      def interval_off
        @interval_seconds = 0
        self
      end

      def build
        parsed = URI.parse(@uri)
        
        if parsed.path && parsed.path != '' && parsed.path != '/'
          raise ArgumentError, "uri must be path-less: #{@uri}"
        end
        if parsed.query
          raise ArgumentError, "uri must be query-less: #{@uri}"
        end

        path = @verify_only ? "/v1/certs/verify-only" : "/v1/certs"
        final_uri = "#{parsed.scheme}://#{parsed.host}:#{parsed.port}#{path}"

        DatCmsManager.new(
          uri: final_uri,
          token: @token,
          interval_seconds: @interval_seconds,
          verify_only: @verify_only
        )
      end
    end
  end
end
